### PR TITLE
Unify url handling with filesources

### LINKED ISSUE #14658: Proposal for unified filesources and URL fetchers
**Proposal**

This is a proposal to extend galaxy filesources with support for a more abstract urlsource, so that all url handling in Galaxy can take advantage of the rich features that filesources offer. This proposal consists of 4 elements:

a. Enable galaxy to have more generic, pluggable protocol handlers as extensions of file sources
b. Unify filesources and other url sources through an appropriate abstract interface
c. Allow all url sources to have an injectable user_context, allowing them to access credentials from user preferences, the vault etc. just as file sources do today.
d. Allow non listable url sources to be excluded from the UI

The reason for filing this issue is to get feedback on this proposal, including commentary on design, potential pitfalls etc. @uwwint has kindly volunteered to take on an initial implementation, and I'm happy to assist with PR review etc. ping @jmchilton

**Background & Rationale**

This proposal is based on a discussion at: https://github.com/galaxyproject/galaxy/pull/13949
In particular, it arose from the fact that file-source based urls (protocol `gxfiles://`) have existing functionality for reading and injecting user credentials, vault credentials, app settings etc. when processing urls. Other url sources, such as http, base64 and with the recent [PR above](https://github.com/galaxyproject/galaxy/pull/13949), DRS - do not have this access at present, as they are not considered file sources. However, some DRS implementations would nevertheless need access to user specific credentials, as appropriate tokens need to be injected per request. In addition, even http urls for example, could benefit from allowing admins to securely inject tokens if the server requires some auth mechanism. Therefore, unifying the handling of these urls through file sources would allow all url sources to be treated uniformly, with full access to contextual information that can be injected appropriately. With recent work that has gone into refactoring url handling, this seems even more tractable as all urls handled by Galaxy go through the `stream_url_to_file` method.

@uwwint has proposed making all filesources generic url handlers. i.e. instead of being a galaxy specific protocol of gxfiles://, it would be a filesource appropriate protocol such as `s3://`. Each file source would have a method to assess its ability to handle a particular url, returning a score based on how well matched it is to the task. The best matching file source will be tasked with handling the file. Consequently, there would be generic protocol handlers for urls such as ssh, ftp, s3, http etc. Filesources appear to be already designed for this possibility through an overridable `get_scheme` method, which defaults to `gxfiles`. This also offers the advantage of making all url handlers administrator configurable, with the ability for admins to introduce custom implementations with additional security restrictions, authentication mechanisms etc.

**Examples:**

```yaml
- type: s3fs
  label: Genome Ark
  id: genomeark
  doc: Access to Genome Ark open data on AWS.
  bucket: genomeark
  anon: true

- type: s3fs
  label: 1000 Genomes
  id: 1000genomes
  doc: Access to the 1000 Genomes Project with human genetic variation, including SNPs, structural variants, and their haplotype context.
  bucket: 1000genomes
  anon: true
```

In this example, given a url like `s3://1000genomes/somefile.fa`, both the above filesources would be potential matches, but since 1000genomes is a more specific match, that filesource would take precedence. In the event of collisions, some possible solutions are
a. To introduce a priority system
b. To raise a configuration error

**Notes:**

1. `gxfiles://` style urls would continue to be supported for backward compatibility

### PR DESCRIPTION
**Pull request by nuwang**
This PR unifies URL handling with filesources. Closes: https://github.com/galaxyproject/galaxy/issues/14658

### Features
* Enables Galaxy to handle urls through a filesource (at present, http(s), base64, ftp and drs are preconfigured, with s3 being admin configurable)
* Enables all url sources to have an injectable user_context. For example, admins can inject credentials for protected DRS servers as in this [example](https://github.com/galaxyproject/galaxy/pull/15497/files#diff-bab7b438e1334bf5754960667d722c0d35f0589aa2e5741dafda5d59df42e25cR5)
    ```yaml
    - type: drs
      id: mydrsserver
      doc: Test drs repository filesource
      http_headers:
        Authorization: |-
          #import base64
          Basic ${base64.b64encode(str.encode(user.preferences['mydrsserver|username'] + ":" + user.preferences['mydrsserver|password'])).decode()}
    ```
* Enables admins to configure the specificity of the match. For example, admins can inject different credentials for specific sites, so that passwordless access to those sites are possible: [example](https://github.com/galaxyproject/galaxy/pull/15497/files#diff-c851d04683c39c45cc33476069aebf0aabda5a7c839d0f50ad39d33318bea624R4)
    ```yaml
    - type: http
      id: test1
      doc: A specific http url handler
      url_regex: "^https?://www.usegalaxy.org/"
      http_headers:
        Authorization: "Bearer ${user.preferences['oidc|bearer_token']}"
    ```
* Enables various DRS access methods to be handled  by existing filesources (e.g. http and s3 access methods are now routed to filesources: [example](https://github.com/galaxyproject/galaxy/pull/15497/files#diff-aa083f3e7e69ae47b4aef7ed8cf709cd27a18b3e78a217bdc05056eb3995f671R74)
* In future, injecting OIDC single-sign-on tokens should be possible by utilizing the work done in: https://github.com/galaxyproject/galaxy/pull/15300

### Backward compatibility
* No changes to the client are necessary.
* Filesources that inherit from BaseFileSource generally do not require any changes. (*except)
* Except that an additional kwargs parameter has been introduced to some method signatures as a means of future expansion and for supporting [chained DRS resolution](https://github.com/galaxyproject/galaxy/pull/15497/files#diff-394e589bbc6b016c487d5373c64a4376c6ded4ebb7633d05487a42ca5ac8e218R105)
* If a plugin directly implements the FileSource interface, it must now implement the `score_url_match` method to indicate its ability to handle a particular url, `get_browsable` method to indicate that it supports the `SupportsBrowsing` interface and the `to_relative_path` method to convert a gxfile url to a relative path. BaseFileSource handles these cases so that most FIleSources do not need any changes.
* The `get_scheme` method has been removed from the interface definition as it is no longer essential, but remains in the `BaseFileSource`.
* The s3.py filesource has been removed, as the s3fs.py filesource does the same thing. The reasons for removing s3.py include its dependencies being less maintained (last updated in 2019) and having less config documentation, in contrast to s3fs.py

## How to test the changes?
(Select all options that apply)
- [X] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [X] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jmchilton** [comment]
I was somewhat skeptical but this is very nice - wonderful job!

The **kwargs getting passed everywhere is very typical Galaxy and feels at home in this codebase and is a thing I've done with my own plugin frameworks in the past. I think we should perhaps try to get away from that though by adding a bit more structure. It looks like there is a limited number of known options with fixed types for the plugins we've defined - could we capture all of those in a TypedDict and then use typing_extensions.Unpack to annotate all those kwargs.

- https://peps.python.org/pep-0692/#using-unpack-with-types-other-than-typeddict
- https://github.com/python/mypy/pull/13471

It looks like TypedDict can specify they don't capture the totality of options (https://mypy.readthedocs.io/en/stable/more_types.html#totality), which would allow downstream plugins to do their own thing as long as they don't conflict with existing plugins.

If you're uninterested in working through all the details - would you be willing to at least defined the dictionaries of possible options for me and annotate what interfaces accept which ones with a comment or something? I could probably work through the possibilities but you're much closer to the code now. Even if we don't figure out how to use the experimental mypy features - I think that would serve as good documentation for the module.

File: test/unit/files/test_drs.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        file_sources = configured_file_sources(FILE_SOURCES_CONF)
+        file_source_pair = file_sources.get_file_source_path(test_url)
+
+        assert file_source_pair.path == test_url
+        assert file_source_pair.file_source.id == "test1"
+
+        assert_realizes_as(file_sources, test_url, "hello drs world", user_context=user_context)
+
+
+@responses.activate
```
Comment (jmchilton) [inline test/unit/files/test_drs.py:67]: So cool!

**Comment by nuwang** [comment]
> I think we should perhaps try to get away from that though by adding a bit more structure. It looks like there is a limited number of known options with fixed types for the plugins we've defined - could we capture all of those in a TypedDict and then use typing_extensions.Unpack to annotate all those kwargs.

This is a very good point and I fully agree with the runaway nature of a kwargs dict - it's hard to stop it from devolving into an incomprehensible bag of data or know who's depending or modifying which bits. I added it because I didn't feel I had many options - it was either a matter of polluting the interface with weird properties, introducing another object similar to a UserContext, or tacking it onto the UserContext itself. Ultimately, I just chose the easy way out with kwargs. I didn't know about TypedDicts so I'm happy to look into that and give it a go, and am also ears to other approaches to handling this.

**Comment by jmchilton** [comment]
Hats off to you @nuwang - this is wonderful!

**Pull request merged**